src/wifipaf/WiFiPAFEndPoint.cpp

```diff
@@ -0,0 +1,1273 @@
+    }
+exit:
+    return err;
+#else
+    return CHIP_NO_ERROR;
+#endif
+}
+
+CHIP_ERROR WiFiPAFEndPoint::Receive(PacketBufferHandle && data)
+{
+    SequenceNumber_t ExpRxNextSeqNum = mPafTP.GetRxNextSeqNum();
+    SequenceNumber_t seqNum;
+    Encoding::LittleEndian::Reader reader(data->Start(), data->DataLength());
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = GetPktSn(reader, data->Start(), seqNum);
+    if (err != CHIP_NO_ERROR)
+    {
+        // Failed to get SeqNum. => Pass down to PAFTP engine directly
+        return _Receive(std::move(data));
+    }
+    /*
+        If reorder-queue is not empty => Need to queue the packet whose SeqNum is the next one at
+        offset 0 to fill the hole.
+    */
+    if ((ExpRxNextSeqNum == seqNum) && (ItemsInReorderQueue == 0))
+        return _Receive(std::move(data));
+
+    ChipLogError(WiFiPAF, "Reorder the packet: [%u, %u]", ExpRxNextSeqNum, seqNum);
+    // Start reordering packets
+    SequenceNumber_t offset = OffsetSeqNum(seqNum, ExpRxNextSeqNum);
+    if (offset >= PAFTP_REORDER_QUEUE_SIZE)
+    {
+        // Offset is too big
+        // => It may be the unexpected packet or duplicate packet => drop it
+        ChipLogError(WiFiPAF, "Offset (%u) is too big => drop the packet", offset);
+        ChipLogDebugBufferWiFiPAFEndPoint(WiFiPAF, data);
+        return CHIP_NO_ERROR;
+    }
+
+    // Save the packet to the reorder-queue
+    if (ReorderQueue[offset] == nullptr)
+    {
+        ReorderQueue[offset] = std::move(data).UnsafeRelease();
+        ItemsInReorderQueue++;
+    }
+
+    // Consume the packets in the reorder queue if no hole exists
+    if (ReorderQueue[0] == nullptr)
+    {
+        // The hole still exists => Can't continue
+        ChipLogError(WiFiPAF, "The hole still exists. Packets in reorder-queue: %u", ItemsInReorderQueue);
+        return CHIP_NO_ERROR;
+    }
+    uint8_t qidx;
+    for (qidx = 0; qidx < PAFTP_REORDER_QUEUE_SIZE; qidx++)
+    {
+        // The head slots should have been filled. => Do rx processing
+        if (ReorderQueue[qidx] == nullptr)
+        {
+            // Stop consuming packets until the hole or no packets
+            break;
+        }
+        // Consume the saved packets
+        ChipLogProgress(WiFiPAF, "Rx processing from the re-order queue [%u]", qidx);
+        err                = _Receive(System::PacketBufferHandle::Adopt(ReorderQueue[qidx]));
+        ReorderQueue[qidx] = nullptr;
+        ItemsInReorderQueue--;
+    }
+    // Has reached the 1st hole in the queue => move the rest items forward
+    // Note: It's to continue => No need to reinit "i"
+    for (uint8_t newId = 0; qidx < PAFTP_REORDER_QUEUE_SIZE; qidx++, newId++)
+    {
+        if (ReorderQueue[qidx] != nullptr)
+        {
+            ReorderQueue[newId] = ReorderQueue[qidx];
+            ReorderQueue[qidx]  = nullptr;
+        }
+    }
+    return err;
+}
+
+CHIP_ERROR WiFiPAFEndPoint::_Receive(PacketBufferHandle && data)
```

Comment:
For all methods, please remove `_` for private methods, it's not needed to have previx.

Reply:
Yes, I will change the name